spark/src/main/java/io/ddf/spark/content/SchemaHandler.java

```diff
@@ -1,386 +1,466 @@
+        throw new DDFException("Error getting factor level counts", e);
+      }
+
+      if (listLevelCounts == null) {
+        throw new DDFException("Error getting factor levels counts");
+      }
+      Map<String, Map<String, Integer>> listLevelCountsWithName = new HashMap<String, Map<String, Integer>>();
+
+      for (Integer columnIndex : columnIndexes) {
+        String colName = this.getDDF().getColumnName(columnIndex);
+        listLevelCountsWithName.put(colName, listLevelCounts.get(columnIndex));
+      }
+      return listLevelCountsWithName;
+    } else {
+      return new HashMap<String, Map<String, Integer>>();
+    }
+  }
+
+  @Override
+  public List<Object> computeFactorLevels(String columnName) throws DDFException {
+    DataFrame df = (DataFrame) this.getDDF().getRepresentationHandler().get(DataFrame.class);
+    DataFrame distinctDF = df.select(columnName).distinct();
+    Long distinctCount = distinctDF.count();
+    if (distinctCount > Factor.getMaxLevelCounts()) {
+      throw new DDFException(String
+          .format("Number of distinct values in column %s is %s larger than MAX_LEVELS_COUNTS = %s", columnName,
+              distinctCount, Factor.getMaxLevelCounts()));
+    }
+
+    List<Object> listValues = FactorIndexer.getFactorMapForColumn(this.getDDF(), columnName).values();
+    return listValues;
+  }
+
+  /**
+   * applySchema(Schema newSchema) will do apply change column type from current to new data type. This function return
+   * new DDF and also statistic when its applied new schema
+   *
+   * @param schema    will use to apply for current DDF
+   * @param isDropRow flag is true, we will drop row which contains any column couldn't parse to new type
+   * @return Tuple2<SparkDDF,ApplySchemaStatistic> +SparkDDF: new sparkddf which applied new schema already
+   * +ApplySchemaStatistic:  applied schema statistic @see ApplySchemaStatistic
+   */
+  public Tuple2<SparkDDF, ApplySchemaStatistic> applySchema(Schema schema, boolean isDropRow) throws DDFException {
+    long startApplySchema = System.currentTimeMillis();
+    // Keep info to Convert column at <pos> to <ColumnType>
+    Tuple2<List<Tuple2<Integer, Column>>, StructType> convertInfo = getConvertInfo(this.getSchema(), schema);
+    List<Tuple2<Integer, Column>> listConvertColumns = convertInfo._1();
+    StructType newSchema =convertInfo._2();
+    try {
+      SparkDDFManager manager = (SparkDDFManager) this.getManager();
+      Accumulator<ApplySchemaStatistic> statisticAccumulator = manager.getSparkContext().accumulator(new ApplySchemaStatistic(), ApplySchemaStatisticParam.getInstance());
+
+      JavaRDD<Row> rdd = ((SparkDDF) this.getDDF()).getRDD(Row.class).toJavaRDD();
+
+
+//      JavaRDD<Row> appliedRdd = rdd.map(row -> convertRowType(row, listConvertColumns, statisticAccumulator, newSchema,isDropRow));
+      JavaRDD<Row> appliedRdd =rdd.map(row -> {
+        ApplySchemaUtils utils = new ApplySchemaUtils();

```

Comment:
this will create new ApplySchemaUtils object for each row in rdd, it's expensive, why don't just make converRowType static method?